core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt

```diff
@@ -1,84 +1,81 @@
 package com.amplitude.core.utilities
 
-import com.amplitude.id.utilities.KeyValueStore
+import com.amplitude.common.Logger
+import com.amplitude.core.events.BaseEvent
 import com.amplitude.id.utilities.createDirectory
 import kotlinx.coroutines.sync.Mutex
 import kotlinx.coroutines.sync.withLock
 import org.json.JSONArray
-import java.io.BufferedReader
+import org.json.JSONException
+import org.json.JSONObject
 import java.io.File
-import java.io.FileOutputStream
-import java.util.Collections
-import java.util.concurrent.ConcurrentHashMap
 
+// Supports 2 event file versions:
+// * version 1:
+//      file path: BASEDIR/STORAGEKEY-INDEX, for example: BASEDIR/$default_instance-0, BASEDIR/$default_instance-1.tmp
+//      internal format: single-line JSON event array
+// * version 2 (current):
+//      file path: BASEDIR/STORAGEKEY/TIMESTAMP-STORAGEID, for example: BASEDIR/$default_instance/1692717940123-abcdefg_hjklmno, BASEDIR/$default_instance/1692718000555-abcdefg_hjklmno.tmp
+//      internal format: each line is a JSON event
 class EventsFileManager(
-    private val directory: File,
+    directory: File,
     private val storageKey: String,
-    private val kvs: KeyValueStore
+    private val logger: Logger,
+    private val getCurrentTimestamp: () -> Long = System::currentTimeMillis
 ) {
+    private val directory = directory.resolve(storageKey)
+    val id = "${ID}_${generateRandomString(INSTANCE_ID_LENGTH)}"
+    private val idRegex = "-$id($|\\.|-)".toRegex()
+    private var lastUsedTimestamp: Long = -1
+
     init {
-        createDirectory(directory)
+        createDirectory(this.directory)
+        attachPreviousV1Files()
+        attachPreviousV2Files()
     }
 
-    private val fileIndexKey = "amplitude.events.file.index.$storageKey"
-
     companion object {
         const val MAX_FILE_SIZE = 975_000 // 975KB
+        private const val ID_LENGTH = 7
+        private const val INSTANCE_ID_LENGTH = 7
+        const val TIMESTAMP_SIZE = 13 // Size of unix timestamp in milliseconds.
+
+        // Static part of storage Id - changed on each app run.
+        val ID = generateRandomString(ID_LENGTH)
+
+        private fun generateRandomString(length: Int): String {
+            val allowedChars = ('a'..'z') + ('0'..'9')
+            return (1..length)
+                .map { allowedChars.random() }
+                .joinToString("")
+        }
     }
 
-    val writeMutex = Mutex()
-    val readMutex = Mutex()
-    val filePathSet: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap<String, Boolean>())
-    val curFile: MutableMap<String, File> = ConcurrentHashMap<String, File>()
+    private val writeMutex = Mutex()
```

Comment:
Thanks @falconandy.

I was hoping that we could resolve the issue by update the mutexs shared a single instance for a given `instanceName` rather than being isolated to each storage instance.

I would like to avoid changing the storage format or file structure if possible.

Some ideas
1. `getMutex(directory, storageKey): Mutex` a provider to create a mutex, or return existing mutex for this directory/storageKey. This could allow multiple file managers to access the same files without concurrency issues.
2. `getEventsFileManager(): EventFileManager` a provider for the EventsFileManager instance used in the FileStorage, that could ensure a single instance of EventsFileManager for a given directory, storageKey. This may be less code, but also less ideal because `logger` and other constructor args would likely need to be a single instance as well.

wdyt?

Reply:
@justin-fiedler 

Current format: single line with json array. If an app crashes/terminates for some reason, the line can be unfinished (an event object can be written partially) - in this case all events will be lost, we can't repair the json array easily. With new format (one line per event) we can skip only broken lines - most lines/events will be correct.

I would prefer timestamps instead of indexes. Indexes are stored in storage - it complicates code.

```
let client1 = Amplitude(Configuration())
let client2 = Amplitude(Configuration())
```

Should `client1.flush()` flush client2's events? What if  flush options are different?

Right now code doesn't support "sharing" properly, there are not-shared local state. A couple of examples:
https://github.com/amplitude/Amplitude-Kotlin/blob/main/core/src/main/java/com/amplitude/core/platform/EventPipeline.kt#L25
https://github.com/amplitude/Amplitude-Kotlin/blob/main/core/src/main/java/com/amplitude/core/platform/intercept/IdentifyInterceptor.kt#L29

Also look at https://github.com/amplitude/Amplitude-Kotlin/blob/main/core/src/main/java/com/amplitude/core/utilities/EventsFileManager.kt#L30-L32 - how long read "locks" are implemented to prevent read the same files.

It looks like we should ensure no Amplitude duplicates with the same instance name.  Maybe `Amplitude.getInstance(instanceName)` is a better way instead of `Amplitude(...)`. Alternatively we can throw/log a error in ``Amplitude(...)` if new instance with existing instance name is created.

wdyt?